src/cpp/session/modules/SessionFiles.cpp

```diff
@@ -1347,6 +1364,281 @@ SEXP rs_pathInfo(SEXP pathSEXP)
    return R_NilValue;
 }
 
+struct ListFilesOptions
+{
+   bool allFiles;
+   bool fullNames;
+   bool recursive;
+   bool includeFiles;
+   bool includeDirs;
+   bool noDotDot;
+};
+
+class ListFilesAcceptAll
+{
+public:
+   bool operator()(const boost::filesystem::path& path)
+   {
+      return true;
+   }
+};
+
+class ListFilesAcceptMatching
+{
+public:
+#ifdef BOOST_WINDOWS_API
+   typedef boost::wregex regex_type;
+   typedef std::wstring string_type;
+#else
+   typedef boost::regex regex_type;
+   typedef std::string string_type;
+#endif
+   
+   ListFilesAcceptMatching(const string_type& pattern, int flags)
+      : pattern_(pattern, flags)
+   {
+   }
+   
+   bool operator()(const boost::filesystem::path& path)
+   {
+      return boost::regex_search(path.native(), pattern_);
+   }
+   
+private:
+   regex_type pattern_;
+};
+
+template <typename F>
+void listFilesImpl(
+      const boost::filesystem::path& path,
+      const boost::filesystem::path& prefix,
+      const ListFilesOptions& options,
+      F&& accept,
+      std::vector<boost::filesystem::path>* pResult)
+{
+   // iterate over other files in the directory
+   try
+   {
+      boost::filesystem::directory_iterator it(path);
+      boost::filesystem::directory_iterator end;
+      for (; it != end; it++)
+      {
+         // skip hidden files if requested
+         auto&& name = it->path().filename();
+         if (!options.allFiles && name.size() > 0 && *name.c_str() == name.dot)
+            continue;
+
+         // construct new prefix
+         auto newPrefix = prefix.empty() ? name : prefix / name;
+         
+         // check if this file is a directory (ignore errors)
+         boost::system::error_code ec;
+         if (boost::filesystem::is_directory(it->status(ec)))
+         {
+            if (options.recursive)
+            {
+               if (options.includeDirs && accept(name))
+                  pResult->push_back(newPrefix);
+               listFilesImpl(it->path(), newPrefix, options, accept, pResult);
+            }
+            else
+            {
+               // ignore options.includeDirs in non-recursive listings
+               if (accept(name))
+                  pResult->push_back(newPrefix);
+            }
+         }
+         else
+         {
+            if (options.includeFiles && accept(name))
+               pResult->push_back(newPrefix);
+         }
+      }
+   }
+   CATCH_UNEXPECTED_EXCEPTION
+   
+}
+
+template <typename F>
+void listFilesDispatch(
+      const std::vector<boost::filesystem::path>& paths,
+      const ListFilesOptions& options,
+      F&& accept,
+      std::vector<boost::filesystem::path>* pResult)
+{
+   // iterate through other files
+   for (auto&& path : paths)
+   {
+      // check for existence (swallow other errors)
+      boost::system::error_code ec;
+      if (boost::filesystem::exists(path, ec))
```

Comment:
Did you have in your tests a test for a non-existent path? I didn't notice it, but it could be helpful?

Reply:
Good idea! I'll add that.